nixos/modules/services/backup/bupstash.nix

```diff
@@ -0,0 +1,294 @@
+                inherit (job) startAt;
+
+                path = [ pkgs.coreutils ] ++ (optional hasSSH pkgs.openssh);
+
+                preStart = ''
+                  mkdir -p send-logs/${name}
+
+                  if [ ! -d '${keysDir}' ]; then
+                    mkdir -p ${keysDir}
+
+                    # Generate new keys
+                    ${bupstash} new-key -o ${keysDir}/master.key
+
+                    ${bupstash} new-sub-key \
+                      -k ${keysDir}/master.key \
+                      -o ${keysDir}/put.key \
+                      --put
+
+                    ${bupstash} new-sub-key \
+                      -k ${keysDir}/master.key \
+                      -o ${keysDir}/list.key \
+                      --list-contents
+
+                  elif [ -f '${keysDir}/master.key' ]; then
+                    echo "[!] The master key for the job '${name}' is present, you might want to store it offline and delete it on the server."
+                  fi
+                '';
+
+                script = builtins.concatStringsSep "\n" (
+                  [ job.preStash ]
+                  ++ (builtins.map
+                    (
+                      r:
+                      let
+                        repo = if builtins.isString r then r else "ssh://${r.user}@${r.server}/${r.path}";
+                        sendlog = builtins.replaceStrings [ "/" ] [ "_" ] (
+                          if builtins.isString r then r else "ssh-${r.server}-${r.path}"
+                        );
+                      in
+                      ''
+                        ${optionalString (builtins.isAttrs r)
+                          "ssh-keygen -F ${r.server} 2>&1 > /dev/null || ssh-keyscan ${r.server} >> ~/.ssh/known_hosts"}
+                        ${bupstash} init -r ${repo} || echo "[-] Repository '${repo}' already exists for '${name}'."
+                        ${bupstash} put ${optionalString (job.command != null) "--exec"} \
+                          --repository "${repo}" \
+                          --key "${keysDir}/put.key" \
+                          --send-log "send-logs/${name}/${sendlog}" \
+                          ${builtins.toString job.extraOptions} \
+                          ${
+                            builtins.toString (builtins.attrValues (builtins.mapAttrs (tag: val: "${tag}='${val}'") job.tags))
+                          } \
+                          :: ${if (job.command == null) then builtins.toString job.paths else job.command}
+                      ''
+                    )
+                    job.repositories
+                  )
+                  ++ [ job.postStash ]
+                );
+
+                serviceConfig = {
+                  Type = "oneshot";
+                  RuntimeDirectory = "bupstash";
+                  StateDirectory = "bupstash";
+                  WorkingDirectory = "/var/lib/bupstash";
+                  User = job.user;
+                  Group = job.group;
```

Comment:
I added `CAP_DAC_READ_SEARCH` to the capabilities, so that the service is able to read all files, but not modify them, optionally, a user and group can be specified